src/tree/param.h

```diff
@@ -429,9 +430,9 @@ template <typename param_t>
     if (constraint == 0) {
       return gain;
     } else if (constraint > 0) {
-      return wleft < wright ? gain : 0.0;
+      return wleft <= wright ? gain : negative_infinity;
     } else {
-      return wleft > wright ? gain : 0.0;
+      return wleft >= wright ? gain : negative_infinity;
```

Comment:
What is a reason for using -Inf here?

Reply:
To ensure that the split never happens.

Reply:
I understand the general idea, but I'm curious why zero is not sufficient now, and whether there might potentially be any corner cases when the negative value might matter.

Reply:
As far as I understand- and I may be completely wrong- the split is calculated on the basis of the change in loss (loss_chg). You can see here;

(https://github.com/dmlc/xgboost/blob/master/src/tree/updater_fast_hist.cc#L869)

that the loss change (at least in the fast histogram updater) is calculated as the gain of the split minus the gain of the current node, hence the loss change can be negative.

Therefore, the negative infinity ensures that the loss change is as small as possible for any split which violates the constraints.

Reply:
My memory could be wrong, but I thought that the negative loss change wasn't allowed by design.
However, if using -Inf doesn't break anything I'm ok with that, except please make the shortcut defined as const.

And I'm not sure whether that java_test fail is real or if travis is being a flaky primadonna again @superbobry ?

Reply:
There's no obvious reason to me why this should break the code.

Okay, that makes sense. I'll set the shortcut to a constant.

Reply:
@redditur were you going to set the const?

Reply:
@redditur Let me argue why the change is not strictly necessary, in terms of algorithmic correctness. The functions `CalcGain` and `CalcGainGivenWeight` always return nonnegative values by design. With algebra we can show that
```
Gain = G^2 / (H + lambda)
```
where G and H are the sums of first- and second-order gradients, respectively. Since all second-order gradients are nonnegative, H is also nonnegative and so is the gain. Since all gains are nonnegative, assigning gain 0 for invalid splits is okay.

At this point, we are arguing about aesthetics rather than correctness. The fact I just described is implicit and not well documented in code. So I'd still include `-Inf` for the sake of clarity.

Reply:
@redditur And maybe I'm nitpicking here, but I'd use
```cpp
return wleft > wright ? gain : negative_infinity
```
instead of
```cpp
return wleft >= wright ? gain : negative_infinity
```
The existing code invalidates the split whenever `wleft == wright`, so we should honor this behavior.